src/ripple/nodestore/impl/Shard.cpp

```diff
@@ -830,6 +869,16 @@ Shard::finalize(
 
         setFileStats(lock);
         lastAccess_ = std::chrono::steady_clock::now();
+
+        if (detShard)
+        {
+            /* Close non-deterministic shard database. */
+            backend_->close();
+            /* Replace non-deterministic shard by deterministic one. */
+            detShard->close();
+            /* Re-open deterministic shard database. */
+            backend_->open(false);
+        }
```

Comment:
There are several issues here.

The DatabaseShard_test unit test fails. The reason being that the SQLite ledger and transaction database handles are open when the DeterministicShard close function tries to remove the directory. Having the DeterministicShard delete the directory of the normal shard isn't optimal.

`backend_->close()` cannot be called as the backend_ may be in use by another thread. That is why the shard class now uses the `Count` class and the `try close` function.

One possible solution...

1. Set an atomic flag to true, that tells `Shard::makeBackendCount` to immediately return `nullptr` (or block until notified).
2. The thread should then wait until notified (via a condition variable) that `backendCount_` equals one since we already have a local `scopedCount` in `finalizeInternal`.
3. Obtain the lock.
4. `backend_->setDeletePath()`, then `backend_->close()`. Reset the `lgrSQLiteDB_`, `txSQLiteDB_` shared pointers. Also `pCache_->reset()`, `nCache_->reset()`, `app_.getShardFamily()->getFullBelowCache(lastSeq_)->reset()`,`app_.getShardFamily()->getTreeNodeCache(lastSeq_)->reset()`
5. `DeterministicShard::close`. Refactor `DeterministicShard::close` to only flush and close its backend.
6. Rename the deterministic NuDB files to the shard file names.
7. Open the shard `Shard::open`.
8. Release the lock.
9. Set the atomic flag to false.

Reply:
Done.